### PR TITLE
Adds a basic Letterhead advertising integration 👋

### PR DESCRIPTION
**Pull request by michaelschofield**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Hey there 👋. This pull requests adds a very simple [Letterhead](https://tryletterhead.com) integration with Newspack, which allows folks to use Letterhead self-service ads in their newsletters. The only interface is a new setting where a user can optionally drop a Letterhead API key. After that, the implementation works the same as Newspack Newsletter's native ad post types, inserting on save.

Closes # .

### How to test the changes in this Pull Request:

1. Ping me for an API key : ) and an account. You'll be able to log in to your account by requesting a new password, once I've generated an admin account for ya.

2. Add the key to the Newsletters Settings field (at the bottom). This will connect you to a dummy account called Black Bitter Coffee I set up at https://app.tryletterhead.com 

3. You can check the promotions or create some more in the app, but all you should need to do in Newspack is send a test email for a day where there is an ad.

4. You should expect to see one or more promotions in your test email. They'd look like:

image

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [3] Have you successfully ran tests with your changes locally?

**Other changes**: I was invited to do a little cleanup here while in the plugin, so I did a little bit of abstraction - turning big functions into smaller functions. I tried to deeply comment everything to help devs navigate.

Edited 9/17 to update some URLs.

### CODE REVIEW

File: includes/service-providers/letterhead/class-newspack-newsletters-letterhead.php [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+	/**
+	 * Pass the json response from, presumably, a Letterhead API call, and return an array of
+	 * Promotions.
+	 *
+	 * @param string $promotions_json The json response from a successful Lettherhead API query.
+	 * @return array Newspack_Newsletters_Letterhead_Promotion[]
+	 */
+	private function get_promotions_from_json_response( $promotions_json ) {
+		$promotion_response_object = json_decode( $promotions_json, false );
```
Comment (adekbadek) [inline includes/service-providers/letterhead/class-newspack-newsletters-letterhead.php:104]: In my testing, I got `Nada.` string as a response (this happened right after opening the editor with a new post). I suppose an empty array was intended?

File: includes/class-newspack-newsletters-settings.php [hunk 2]
```diff
@@ -97,6 +97,22 @@ public static function get_settings_list() {
...
+			 * API.
+			 *
+			 * @see https://help.tryletterhead.com/hc/en-us/articles/360051316834-API-reference
+			 */
+			array(
+				'default'     => '',
+				'description' => __( 'Letterhead Key (optional)', 'newspack-newsletters' ),
+				'key'         => Newspack_Newsletters_Letterhead::LETTERHEAD_WP_OPTION_KEY,
+				'type'        => 'text',
+			),
```
Comment (adekbadek) [inline includes/class-newspack-newsletters-settings.php:115]: It'd be nice to add a short explanation text with a signup link. This would help discoverability and also clarify what's that setting about.

**Review by adekbadek — CHANGES_REQUESTED** [review]
After inserting the API key creating a new newsletter with a date matching the date of the Letterhead promotion I did receive an email with the ad creative, as outlined in the description. 
There are a couple of issues I've pointed out, and one more - the plugin displays a pre-send modal with information about the ads. When Letterhead ads are active, this should be reflected in the modal:
image

And thanks for the refactor!

**Comment by michaelschofield** [comment]
@adekbadek Hey there 👋. Thanks for the patience. We didn't have enough bandwidth to allocate in this direction since your review --_-- but **now we do**. This is one of my priorities this and next sprint, so you should hear form me in a normal human way to make up for being such a ghost 👻

File: includes/class-newspack-newsletters-renderer.php [hunk 3]
```diff
@@ -637,39 +783,37 @@ private static function insert_ads( $markup, $current_position ) {
...
 		$valid_blocks = array_filter(
-			parse_blocks( $post->post_content ),
+			parse_blocks( $content ),
 			function ( $block ) {
 				return null !== $block['blockName'];
 			}
 		);
+
+		/**
+		 * The total length of the Newsletter, which we use to hashtag math and determine
```
[the remark this answers is not in the document]
Comment (michaelschofield) [reply includes/class-newspack-newsletters-renderer.php:805]: Haha. Sorry about these.

File: includes/class-newspack-newsletters-renderer.php [hunk 4]
```diff
@@ -698,7 +842,7 @@ function ( $block ) {
 				// Insert ads between other blocks.
 				$block_content = self::render_mjml_component( $block );
 				if ( $include_ads ) {
-					$current_position += strlen( wp_strip_all_tags( $block_content ) );
```
[the remark this answers is not in the document]
Comment (michaelschofield) [reply includes/class-newspack-newsletters-renderer.php:701]: Honestly _I'm not sure_. This might be a casualty of multitasking. Good eye. I'll re-implement (and/or find a *really good reason I did such-and-such and let you know*). But off the top of my head I think this was some scorched-earth.

File: includes/service-providers/letterhead/class-newspack-newsletters-letterhead.php (hunk 1 above)
Comment (michaelschofield) [reply includes/service-providers/letterhead/class-newspack-newsletters-letterhead.php:104]: image

When I make the next pull maybe we can coordinate and you can show me if it fails again : ). After merging master down and making some of the changes you suggested, I was able to load a promotion - it might have been that the account you were pointing to just didn't have a promotion set for that date, but there's a lot of reasons it could turn up goose-eggs : ). If there aren't any promotions on that day it will return an empty array, yup.

File: includes/class-newspack-newsletters-settings.php (hunk 2 above)
Comment (michaelschofield) [reply includes/class-newspack-newsletters-settings.php:115]: Iiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiii _linked up_ "Letterhead" in my next pull, but I wasn't totally sure how I might add an additional description to this array. We have some Newspack users who are also Letterhead users already and I think that's just via some partnership and some demoing that's happened already.

File: includes/service-providers/letterhead/class-newspack-newsletters-letterhead.php (hunk 1 above)
Comment (adekbadek) [reply includes/service-providers/letterhead/class-newspack-newsletters-letterhead.php:104]: > If there aren't any promotions on that day it will return an empty array, yup.

Hm, I'm still getting a `Nada.` response with no upcoming promotions :/

File: includes/class-newspack-newsletters-renderer.php [hunk 5]
```diff
@@ -760,9 +955,35 @@ public static function mjml_api_credentials() {
...
+	 */
+	public function register_newsletter_renderer_routes() {
+		$arguments = array(
+			/**
+			 * Return an array of properties required to render a useful ads warning.
+			 *
+			 * @uses Newspack_Newsletters_Renderer::get_ads_warning_in_editor()
+			 */
+			'callback' => array( $this, 'get_ads_warning_in_editor' ),
+			'methods'  => 'GET',
```
Comment (adekbadek) [inline includes/class-newspack-newsletters-renderer.php:971]: A `permission_callback` key [is required in arguments since WP 5.5](https://developer.wordpress.org/reference/functions/register_rest_route/#changelog)

File: includes/class-newspack-newsletters-renderer.php [hunk 6]
```diff
@@ -637,42 +833,37 @@ private static function insert_ads( $markup, $current_position ) {
...
+		 * The total length of the Newsletter, which we use to determine
+		 * the position an ad is included.
+		 */
 		$total_length = self::get_total_newsletter_character_length( $valid_blocks );
 
-		// Gather ads.
+		/**
+		 * When ads are enabled, we fetch and format them for insertion.
+		 *
+		 * @note "diable_ads" is a typo that's been in production for awhile.
```
Comment (adekbadek) [inline includes/class-newspack-newsletters-renderer.php:863]: 😬😅

File: composer.json [hunk 7]
```diff
@@ -5,7 +5,8 @@
   "require": {
     "drewm/mailchimp-api": "^2.5",
     "constantcontact/constantcontact": "2.1.*",
-    "campaignmonitor/createsend-php": "^6.1"
+    "campaignmonitor/createsend-php": "^6.1",
+    "xwp/wp-dev-lib": "^1.6"
```
Comment (adekbadek) [inline composer.json:9]: I can't see that this library is used somewhere 🤔

File: src/components/send-button/index.js [hunk 8]
```diff
@@ -104,22 +105,28 @@ export default compose( [
...
+				if ( countOfActiveAds > 0 ) {
 					setAdsWarning(
-						sprintf(
+						sprintf( /* eslint-disable-line */
 							_n(
-								'There is %d active ad.',
-								'There are %d active ads.',
-								activeAds.length,
+								'There is %d active ' + response.label + '.',
+								'There is %d active ' + response.label + 's.',
```
Comment (adekbadek) [inline src/components/send-button/index.js:125]:
```suggestion
								'There are %d active ' + response.label + 's.',
```
?

File: includes/class-newspack-newsletters-renderer.php [hunk 9]
```diff
@@ -760,9 +955,35 @@ public static function mjml_api_credentials() {
...
+			 */
+			'callback' => array( $this, 'get_ads_warning_in_editor' ),
+			'methods'  => 'GET',
+		);
+
+		$ad_type_slug = Newspack_Newsletters_Ads::NEWSPACK_NEWSLETTERS_ADS_CPT;
+		$namespace    = "wp/v2/{$ad_type_slug}";
+		$route        = 'count';
+
+		return register_rest_route( $namespace, $route, $arguments );
```
Comment (adekbadek) [inline includes/class-newspack-newsletters-renderer.php:978]: Registering a REST route in the renderer feels weird, I'd suggest registering the new route among the other REST routes, in `Newspack_Newsletters::rest_api_init`.

**Review by adekbadek — CHANGES_REQUESTED** [review]
Some other remarks:
1. `package-lock.json` is updated even though there are no changes to `package.json` - this can probably be omitted 
2. Some things changed on `master` in the meantime - most notably #443, this PR should be updated

File: includes/class-newspack-newsletters-renderer.php (hunk 6 above)
Comment (michaelschofield) [reply includes/class-newspack-newsletters-renderer.php:863]: Lol

File: composer.json (hunk 7 above)
Comment (michaelschofield) [reply composer.json:9]: _Oh_ this one I can remove i think. The Automattic / PHP automagic commit lint thing was referring to some lint rules defined in this library, which just didn't exist on my end. So I pulled it down to get passed the code sniffer. I can delete this if ya'll don't think you need it.

File: includes/class-newspack-newsletters-renderer.php (hunk 9 above)
Comment (michaelschofield) [reply includes/class-newspack-newsletters-renderer.php:978]: Oh good suggestion. I didn't see that action.

File: includes/service-providers/letterhead/class-newspack-newsletters-letterhead.php (hunk 1 above)
Comment (michaelschofield) [reply includes/service-providers/letterhead/class-newspack-newsletters-letterhead.php:104]: I'll touch base with ya and make sure we're hitting the right endpoint.

File: composer.json (hunk 7 above)
Comment (adekbadek) [reply composer.json:9]: Weird, it is used in the pre-commit hook in our [other project/s](https://github.com/Automattic/newspack-plugin/blob/98ee47bdee1222b3985bfde9f5f9f20021d088a8/composer.json#L33), but not in this one. I think it can be deleted, unless I'm missing something here.

**Comment by michaelschofield** [comment]
Good lord these merge conflicts. 😪

**Comment by michaelschofield** [comment]
@adekbadek 👋 Re-requesting your weathered eye. Corrected some merge conflicts since the latest pull, made the changes you requested. I also think I know why you're getting no promotions in your tests. 

In Slack ,remind me to make sure you have the right _key_ to add to your Newsletters -> Settings (just to double check), but otherwise we want to make sure the API is pointed to `platform.staging.whereby.us` (we can define it for a local environment). Then, if you don't want to go setup promotions yourself again, there's one for "April 13, 2021," if you set your newsletter's publication date to that. 

Anyway, if you keep coming up cold turkey, "nada" is the right response in the even that the API found nothing, so we'll just need to doublecheck that a promo is scheduled for the date of a newsletter.

**Review by adekbadek — CHANGES_REQUESTED** [review]
Awaiting access to the API, but I'll post a review meantime. 

> Anyway, if you keep coming up cold turkey, "nada" is the right response in the even that the API found nothing, so we'll just need to doublecheck that a promo is scheduled for the date of a newsletter.

Then that should be handled - if response is `Nada.` (or not an array, anyway), then `$promotion_response_object` should be turned into an empty array [here](https://github.com/Automattic/newspack-newsletters/blob/75947e4eff70d4ac9f20e6fff3b4b8f600620298/includes/service-providers/letterhead/class-newspack-newsletters-letterhead.php#L104).

File: includes/class-newspack-newsletters-renderer.php [hunk 10]
```diff
@@ -657,33 +829,13 @@ public static function post_to_mjml_components( $post, $include_ads = false ) {
...
-					$percentage            = intval( get_post_meta( $ad->ID, 'position_in_content', true ) ) / 100;
-					self::$ads_to_insert[] = [
-						'precise_position' => $total_length * $percentage,
-						'percentage'       => $percentage,
-						'markup'           => self::post_to_mjml_components( $ad ),
-						'is_inserted'      => false,
-					];
-				}
-			}
+			self::$ads_to_insert = self::get_ads( $post->post_date, $total_length );
```
[the remark this answers is not in the document]
Comment (michaelschofield) [reply includes/class-newspack-newsletters-renderer.php:838]: Gotcha

File: includes/class-newspack-newsletters.php [hunk 11]
```diff
@@ -1144,6 +1173,32 @@ public static function debug_mode() {
...
 
+
+	/**
+	 * Get properties required to render a useful modal in the editor that alerts
+	 * users of ads they're sending.
+	 *
+	 * @param WP_REST_REQUEST $request The WP Request Object.
+	 * @return array
+	 */
+	public function get_ads_warning_in_editor( $request ) {
```
[the remark this answers is not in the document]
Comment (michaelschofield) [reply includes/class-newspack-newsletters.php:1184]: @adekbadek Ah. In wordpress 

```
				'callback'            => [ __CLASS__, 'get_ads_warning_in_editor' ],
```

is a static call? I thought it was the equivalent of $this->myMethodHere();.

File: includes/class-newspack-newsletters.php [hunk 12]
```diff
@@ -79,8 +79,15 @@ public function __construct() {
...
 		add_action( 'admin_enqueue_scripts', [ __CLASS__, 'branding_scripts' ] );
 		add_filter( 'newspack_theme_featured_image_post_types', [ __CLASS__, 'support_featured_image_options' ] );
+
+		/**
+		 * Set the email service provider, which will instantiate the appropriate class
+		 * (Mailchimp, ConstantContact, etc.).
+		 */
 		self::set_service_provider( self::service_provider() );
 
+		$renderer = new Newspack_Newsletters_Renderer();
```
[the remark this answers is not in the document]
Comment (michaelschofield) [reply includes/class-newspack-newsletters.php]: Great catch. My IDE even screamed at me for it and missed it. Resolved.

File: src/components/send-button/index.js [hunk 13]
```diff
@@ -104,22 +105,28 @@ export default compose( [
...
-				const activeAds = response.filter( isAdActive );
-				if ( activeAds.length ) {
+				const countOfActiveAds = response.count;
+				setActiveAdManageUrl( response.manageUrl );
+				setAdLabel( response.label );
+
+				if ( countOfActiveAds > 0 ) {
 					setAdsWarning(
-						sprintf(
+						sprintf( /* eslint-disable-line */
```
[the remark this answers is not in the document]
Comment (michaelschofield) [reply src/components/send-button/index.js:122]: Oh woops good catch. I had that there while developing. Removing.

[the remark this answers is not in the document]
Comment (michaelschofield) [reply src/components/send-button/index.js:122]: Ah, I couldn't get accommodate [this rule](https://github.com/WordPress/gutenberg/blob/trunk/packages/eslint-plugin/docs/rules/valid-sprintf.md) in a sane way, mainly because I'm not familiar with _n or _nx. I think I fixed that but then prettier doesn't like it : ). So I didn't remove it this round.

File: includes/class-newspack-newsletters.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (michaelschofield) [reply includes/class-newspack-newsletters.php]: Resolved

**Comment by michaelschofield** [comment]
@adekbadek Hey there. Made some changes and re-requested your review. Regarding

> Then that should be handled - if response is Nada. (or not an array, anyway), then $promotion_response_object should be turned into an empty array [here](https://github.com/Automattic/newspack-newsletters/blob/75947e4eff70d4ac9f20e6fff3b4b8f600620298/includes/service-providers/letterhead/class-newspack-newsletters-letterhead.php#L104).

Good catch. I'm also going to change this API on my end to ensure it's returning consistently. Patched on line 104.

File: includes/class-newspack-newsletters.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/class-newspack-newsletters.php]: Turns out it's static, but I only know this because of a PHP notice. It's useful to `tail` PHP error log while developing.

File: includes/class-newspack-newsletters-renderer.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/class-newspack-newsletters-renderer.php:838]: Please merge in `master` so that's resolved ;)

File: src/components/send-button/index.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (adekbadek) [reply src/components/send-button/index.js:122]: I see now - there's actually an issue here and the rule should not be disabled. `_n` expects static strings as input, as this will be the content for translation - there can't be any interpolated variables there. So it should be:

```js 
sprintf(
	_n(
		'There is %d active %s.',
		'There are %d active %ss.',
		countOfActiveAds,
		'newspack-newsletters'
	),
	countOfActiveAds,
	response.label
)
```

**Review by adekbadek — CHANGES_REQUESTED** [review]
Responded in comment. 

> I'm also going to change this API on my end to ensure it's returning consistently.

Awaiting API change so the no-promotions case can be tested.

**Comment by adekbadek** [comment]
Confirmed that the API now returns an empty array if there are no promotions. Awaiting [changes requested](https://github.com/Automattic/newspack-newsletters/pull/399#pullrequestreview-653207913).

File: includes/class-newspack-newsletters-renderer.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (michaelschofield) [reply includes/class-newspack-newsletters-renderer.php:838]: > Just noting here: the post_date will not be available due to a bug, fixed in #459

@adekbadek 👋 When I look at the [diff you linked to](https://github.com/Automattic/newspack-newsletters/pull/459) it doesn't look like it affects where this method's called. I just confirmed doing a `var_dump($post)` that the whole WP_Post class is there. Things seeeeeeeeeeeem to be a-okay, can you let me know if I'm missing something?

File: src/components/send-button/index.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (michaelschofield) [reply src/components/send-button/index.js]: Fixed 👍

File: includes/class-newspack-newsletters-renderer.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (adekbadek) [reply includes/class-newspack-newsletters-renderer.php:838]: Yep, looks good.

**Review by adekbadek — CHANGES_REQUESTED** [review]
There's a git conflict marker at the bottom of `package-lock.json`, which [prevents `npm` from installing dependencies](https://app.circleci.com/pipelines/github/Automattic/newspack-newsletters/1178/workflows/399e493d-70ca-4cb3-9b2f-bbd9bb644201/jobs/2579). 
The changes in `package.json` seem unrelated to this PR. I suggest removing changes to `package.json` & `package-lock.json` from this PR.

**Comment by michaelschofield** [comment]
Good suggestion 🤘.

**Comment by michaelschofield** [comment]
FYI merged latest master into this master.

**Comment by michaelschofield** [comment]
@adekbadek Alrighty. Looks like we're passing tests and I was able to see a promotion in a test email. Attached a screenshot below. Give it a spin when you have an opportunity: you should see the exact same promotion as in the screenshot provided your newsletter is set for October 8, 2021. Thanks!  
image

File: composer.lock [hunk 14]
```diff
@@ -1517,5 +1517,5 @@
     "prefer-lowest": false,
     "platform": [],
     "platform-dev": [],
-    "plugin-api-version": "2.1.0"
+    "plugin-api-version": "2.0.0"
```
Comment (adekbadek) [inline composer.lock:1520]: Is this change intentional? There are no changes to `composer.json` in this PR and this is an auto-generated file.

File: src/components/send-button/index.js [hunk 15]
```diff
@@ -117,31 +118,39 @@ export default compose( [
...
+		const [ adLabel, setAdLabel ] = useState();
+		const [ adsWarning, setAdsWarning ] = useState();
+		const [ activeAdManageUrl, setActiveAdManageUrl ] = useState();
+		const newspack_ad_post_type_slug = NEWSLETTER_AD_CPT_SLUG;
+
 		let modalSubmitLabel;
 		if ( 'manual' === serviceProviderName ) {
 			modalSubmitLabel = is_public
-				? __( 'Mark as sent and publish', 'newspack-newsletters' )
+				? __( 'Mark as send and publish', 'newspack-newsletters' )
```
Comment (adekbadek) [inline src/components/send-button/index.js:129]: "sent" was intentional - in manual mode, the newsletter is not actually sent (that's for the user to do), so all that's available is marking as sent.

File: src/components/send-button/index.js [hunk 16]
```diff
@@ -217,10 +226,12 @@ export default compose( [
...
 								{ adsWarning }{' '}
-								<a
-									href={ `/wp-admin/edit.php?post_type=${ NEWSLETTER_CPT_SLUG }&page=newspack-newsletters-ads-admin` }
+								<a // eslint-disable-line react/jsx-no-target-blank
+									href={ activeAdManageUrl }
+									rel={ adLabel === 'promotion' ? 'noreferrer' : '' }
+									target={ adLabel === 'promotion' ? '_blank' : '_self' }
 								>
-									{ __( 'Manage ads', 'newspack-newsletters' ) }
+									{ __( 'Manage ' + adLabel + 's.', 'newspack-newsletters' ) }
```
Comment (adekbadek) [inline src/components/send-button/index.js:234]: The first argument should be a string literal, not a computed value. `sprintf` should be used:
```suggestion
									{ sprintf( __( 'Manage %ss.', 'newspack-newsletters' ), adLabel ) }
```

File: includes/class-newspack-newsletters.php [hunk 17]
```diff
@@ -1057,6 +1084,32 @@ public static function debug_mode() {
...
+		$post_date                  = $request->get_param( 'date' );
+		$newspack_ad_type           = Newspack_Newsletters_Ads::NEWSPACK_NEWSLETTERS_ADS_CPT;
+		$url_to_manage_promotions   = 'https://app.tryletterhead.com/promotions';
+		$url_to_manage_newspack_ads = "/wp-admin/edit.php?post_type={$newspack_ad_type}&page=newspack-newsletters-ads-admin";
+
+		$ads = Newspack_Newsletters_Renderer::get_ads( $post_date, 0 );
+
+		return [
+			'count'     => count( $ads ),
+			'label'     => $has_letterhead_credentials ? 'promotion' : 'ad',
```
Comment (adekbadek) [inline includes/class-newspack-newsletters.php:1107]: These two strings should be translatable (wrapped in `__()`). Ideally, the label for the ads modal should be returned from the backend, so there's no need for the `adLabel === 'promotion'` logic.

**Review by adekbadek — CHANGES_REQUESTED** [review]
I confirm that I see a promotion when newsletter is scheduled for the same date as the promotion.

File: includes/class-newspack-newsletters.php (hunk 17 above)
Comment (michaelschofield) [reply includes/class-newspack-newsletters.php:1107]: Good idea.

File: src/components/send-button/index.js (hunk 15 above)
Comment (michaelschofield) [reply src/components/send-button/index.js:129]: Huh. I honestly don't remember making this change (or why I would). Will correct with a github suggestion.
```suggestion
				? __( 'Mark as sent and publish', 'newspack-newsletters' )
```

File: composer.lock (hunk 14 above)
Comment (michaelschofield) [reply composer.lock:1520]: @adekbadek Nope. It might be result of how the requirement is written and running a `composer update`. If you prefer I can checkout the lock file, but I don't know what this library even does so I didn't intentionally update it. Lmk.

Comment (adekbadek) [reply composer.lock:1520]: Let's skip it for this PR.

File: includes/class-newspack-newsletters.php [hunk 18]
```diff
@@ -1057,6 +1084,39 @@ public static function debug_mode() {
...
+		$ads_manage_url_rel    = $has_letterhead_credentials ? 'noreferrer' : '';
+		$ads_manage_url_target = $has_letterhead_credentials ? '_blank' : '_self';
+
+		return [
+			'count'           => count( $ads ),
+			'label'           => $ads_label,
+			'manageUrl'       => $ads_manage_url,
+			'manageUrlRel'    => $ads_manage_url_rel,
+			'manageUrlTarget' => $ads_manage_url_target,
+		];
```
Comment (michaelschofield) [inline includes/class-newspack-newsletters.php:1116]: @adekbadek Hey there. Just wanted to point out the changes I've made in this most recent push to move these values into the API response. The "rel" and the "target" are ultimately for the `<a />`, largely because managing an ad through Letterhead at least right now I think users would prefer to pop open a separate tab (`target="_blank"`) rather than interrupt their current work. The JS linter then requires a `rel="noreferrer"` for external links. 

But this way we can just grab these values from the API and limit what the js is actually responsible for displaying.

File: includes/class-newspack-newsletters.php [hunk 19]
```diff
@@ -1057,6 +1084,39 @@ public static function debug_mode() {
...
+	 * @return array
+	 */
+	public static function get_ads_warning_in_editor( $request ) {
+		$letterhead                 = new Newspack_Newsletters_Letterhead();
+		$has_letterhead_credentials = $letterhead->has_api_credentials();
+		$post_date                  = $request->get_param( 'date' );
+		$newspack_ad_type           = Newspack_Newsletters_Ads::NEWSPACK_NEWSLETTERS_ADS_CPT;
+
+		$url_to_manage_promotions   = 'https://app.tryletterhead.com/promotions';
+		$url_to_manage_newspack_ads = "/wp-admin/edit.php?post_type={$newspack_ad_type}&page=newspack-newsletters-ads-admin";
```
Comment (adekbadek) [inline includes/class-newspack-newsletters.php:1102]:
```suggestion
		$url_to_manage_newspack_ads = "/wp-admin/edit.php?post_type={$newspack_ad_type}";
```

I know this is cut-and-pasted, but for some reason the `page` param breaks the URL.

File: src/components/send-button/index.js [hunk 20]
```diff
@@ -117,6 +118,13 @@ export default compose( [
 			publishStatus = 'publish';
 		}
 
+		const [ adLabel, setAdLabel ] = useState();
+		const [ adsWarning, setAdsWarning ] = useState();
+		const [ activeAdManageUrl, setActiveAdManageUrl ] = useState();
+		const [ activeAdManageUrlRel, setActiveAdManageUrlRel ] = useState();
+		const [ activeAdManageTarget, setActiveAdManageTarget ] = useState();
+		const newspack_ad_post_type_slug = NEWSLETTER_AD_CPT_SLUG;
```
Comment (adekbadek) [inline src/components/send-button/index.js:126]: No need to re-assign a constant to another constant, `NEWSLETTER_AD_CPT_SLUG` can be used.

File: includes/class-newspack-newsletters.php [hunk 21]
```diff
@@ -82,6 +82,11 @@ public function __construct() {
 		add_action( 'save_post_' . self::NEWSPACK_NEWSLETTERS_CPT, [ __CLASS__, 'save' ], 10, 3 );
 		add_action( 'admin_enqueue_scripts', [ __CLASS__, 'branding_scripts' ] );
 		add_filter( 'newspack_theme_featured_image_post_types', [ __CLASS__, 'support_featured_image_options' ] );
+
+		/**
+		 * Set the email service provider, which will instantiate the appropriate class
+		 * (Mailchimp, ConstantContact, etc.).
+		 */
```
Comment (adekbadek) [inline includes/class-newspack-newsletters.php:89]: Probably due to a `master` merge, this comment moved one line up - it should be above the line _after_ the next one.

**Comment by michaelschofield** [comment]
@adekbadek Hey there Adam. Alrighty I think this should do it. Would you look in to the Build CircleCI Build failure for me? It looks like it's failing because it can't find the Python executable - which I think may be on its end, but just confirm and lmk if I need to do anything. Happy to lend a hand.

**Comment by adekbadek** [comment]
Hi @michaelschofield !

> Would you look in to the Build CircleCI Build failure for me? 

This should be fixed by https://github.com/Automattic/newspack-newsletters/pull/625, can you merge in the latest `master`?

**Comment by michaelschofield** [comment]
@adekbadek On it! Thanks!

**Comment by michaelschofield** [comment]
@adekbadek Hot dang. Green checks across the board. Alrighty, ready for your eyes one more time.

**Comment by yogeshbeniwal** [comment]
@michaelschofield Which ESP it will use to send newsletter? Is it what configured with newspack newsletter or configured in Letterhead dashboard?

**Comment by adekbadek** [comment]
@yogeshbeniwal – this will use the ESP configured with Newspack Newsletters.

**Review by adekbadek — APPROVED** [review]

**Comment by davisshaver** [comment]
🥳  Excited to try this!
